common/tasks/interleaved_weighted_round_robin.go

```diff
@@ -0,0 +1,269 @@
+	priorityToWeight map[int]int,
+	processor Processor,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) *InterleavedWeightedRoundRobinScheduler {
+	return &InterleavedWeightedRoundRobinScheduler{
+		status: common.DaemonStatusInitialized,
+		option: option,
+
+		processor:    processor,
+		metricsScope: metricsClient.Scope(metrics.TaskSchedulerScope),
+		logger:       logger,
+
+		notifyChan:   make(chan struct{}, 1),
+		shutdownChan: make(chan struct{}),
+
+		priorityToWeight:     priorityToWeight,
+		weightToTaskChannels: make(map[int]*WeightedChannel),
+		iwrrChannels:         []*WeightedChannel{},
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	s.processor.Start()
+
+	go s.eventLoop()
+
+	s.logger.Info("interleaved weighted round robin task scheduler started")
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	close(s.shutdownChan)
+
+	s.processor.Stop()
+
+	s.rescheduleTasks()
+
+	s.logger.Info("interleaved weighted round robin task scheduler stopped")
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Submit(
+	task PriorityTask,
+
+) {
+	channel := s.getOrCreateTaskChannel(s.priorityToWeight[task.GetPriority()])
+	channel.Chan() <- task
+	s.notifyDispatcher()
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) eventLoop() {
+	for {
+		select {
+		case <-s.notifyChan:
+			s.dispatchTasks()
+		case <-s.shutdownChan:
+			return
+		}
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) getOrCreateTaskChannel(
+	weight int,
+) *WeightedChannel {
+	s.RLock()
+	channel, ok := s.weightToTaskChannels[weight]
+	if ok {
+		s.RUnlock()
+		return channel
+	}
+	s.RUnlock()
+
+	s.Lock()
+	defer s.Unlock()
+
+	channel, ok = s.weightToTaskChannels[weight]
+	if ok {
+		return channel
+	}
+
+	channel = NewWeightedChannel(weight, WeightedChannelDefaultSize)
+	s.weightToTaskChannels[weight] = channel
+
+	weightedChannels := make(WeightedChannels, 0, len(s.weightToTaskChannels))
+	for _, weightedChan := range s.weightToTaskChannels {
+		weightedChannels = append(weightedChannels, weightedChan)
+	}
+	sort.Sort(weightedChannels)
+
+	iwrrChannels := make([]*WeightedChannel, 0, len(weightedChannels))
+	maxWeight := weightedChannels[len(weightedChannels)-1].Weight()
+	for round := maxWeight - 1; round > -1; round-- {
+		for index := len(weightedChannels) - 1; index > -1 && weightedChannels[index].Weight() > round; index-- {
+			iwrrChannels = append(iwrrChannels, weightedChannels[index])
+		}
+	}
+	s.iwrrChannels = iwrrChannels
+
+	return channel
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) dispatchTasks() {
+	for s.hasRemainingTasks() {
+		weightedChannels := s.channels()
+		s.doDispatchTasks(weightedChannels)
+	}
```

Comment:
do you need to check shutdownChan here?

Reply:
the stop function will actively clear tasks by calling `reschedule`